lib/sparse/src/index/posting_list.rs

```diff
@@ -24,6 +41,70 @@ impl PostingList {
         }
         posting_list.build()
     }
+
+    /// Upsert a posting element into the posting list.
+    ///
+    /// Worst case is adding a new element at the end of the list with a very large weight.
+    /// This forces to propagate it as potential max_next_weight to all the previous elements.
+    pub fn upsert(&mut self, mut posting_element: PostingElement) {
+        // find insertion point in sorted posting list
+        let index = self
+            .elements
+            .binary_search_by_key(&posting_element.record_id, |e| e.record_id);
+
+        let modified_index = match index {
+            Ok(found_index) => {
+                // Update existing element for the same id
+                let element = &mut self.elements[found_index];
+                if element.weight == posting_element.weight {
+                    // no need to update anything
+                    return;
```

Comment:
Just a niggle, can we remove `return` to keep functional styling here?

Reply:
What would you prefer instead to shortcut the computation?

Reply:
`modified_index` may be optional. It highlights that that `propagate_max_next_weight_to_the_left` can be skipped

Reply:
Got it, makes sense :+1: 

https://github.com/qdrant/qdrant/pull/2826/commits/4e6e3e89f6e0c1e0bd222f6801ca8512385a9461